tools/gfx/cuda/render-cuda.cpp

```diff
@@ -0,0 +1,1506 @@
+            if (!typeLayout->getType())
+            {
+                if (auto elementTypeLayout = typeLayout->getElementTypeLayout())
+                    typeLayout = elementTypeLayout;
+            }
+
+            switch (typeLayout->getKind())
+            {
+            default:
+                return typeLayout;
+
+            case slang::TypeReflection::Kind::ConstantBuffer:
+            case slang::TypeReflection::Kind::ParameterBlock:
+                typeLayout = typeLayout->getElementTypeLayout();
+                continue;
+            }
+        }
+    }
+
+    CUDAShaderObjectLayout(slang::TypeLayoutReflection* layout)
+    {
+        typeLayout = unwrapParameterGroups(layout);
+
+        // Compute the binding ranges that are used to store
+        // the logical contents of the object in memory. These will relate
+        // to the descriptor ranges in the various sets, but not always
+        // in a one-to-one fashion.
+
+        SlangInt bindingRangeCount = typeLayout->getBindingRangeCount();
+        for (SlangInt r = 0; r < bindingRangeCount; ++r)
+        {
+            slang::BindingType slangBindingType = typeLayout->getBindingRangeType(r);
+            SlangInt count = typeLayout->getBindingRangeBindingCount(r);
+            slang::TypeLayoutReflection* slangLeafTypeLayout =
+                typeLayout->getBindingRangeLeafTypeLayout(r);
+
+            SlangInt descriptorSetIndex = typeLayout->getBindingRangeDescriptorSetIndex(r);
+            SlangInt rangeIndexInDescriptorSet =
+                typeLayout->getBindingRangeFirstDescriptorRangeIndex(r);
+
+            auto uniformOffset = typeLayout->getDescriptorSetDescriptorRangeIndexOffset(
+                descriptorSetIndex, rangeIndexInDescriptorSet);
+
+            Index baseIndex = 0;
+            switch (slangBindingType)
+            {
+            case slang::BindingType::ConstantBuffer:
+            case slang::BindingType::ParameterBlock:
+            case slang::BindingType::ExistentialValue:
+                baseIndex = m_subObjectCount;
+                m_subObjectCount += count;
+                break;
+
+            case slang::BindingType::Sampler:
+                baseIndex = m_samplerCount;
+                m_samplerCount += count;
+                break;
+
+            case slang::BindingType::CombinedTextureSampler:
+                baseIndex = m_combinedTextureSamplerCount;
+                m_combinedTextureSamplerCount += count;
+                break;
+
+            case slang::BindingType::VaryingInput:
+                baseIndex = m_varyingInputCount;
+                m_varyingInputCount += count;
+                break;
+
+            case slang::BindingType::VaryingOutput:
+                baseIndex = m_varyingOutputCount;
+                m_varyingOutputCount += count;
+                break;
+
+            default:
+                baseIndex = m_resourceViewCount;
+                m_resourceViewCount += count;
+                break;
```

Comment:
The rationale for this kind of `switch` statement in a `ShaderObjectLayout` implementation is that it handles the N-to-M mapping from the N `slang::BindingType` cases that binding ranges use in Slang reflection data over to the M distinguished cases that a particular API cares about.

E.g., for a pure D3D11 implementation, the main cases would be CBV, SRV, UAV, and Sampler. Vulkan would have a much higher number of cases, since it recognizes a greater number of cases for descriptor types.

In the CUDA case, a lot of these binding types are either expected to never occur (e.g., `VaryingInput`) or don't really matter when they do (`Sampler`).